### ISSUE TITLE
Possible issues with CPU usage while polling infrequently?

### ISSUE THREAD

**Issue by rrnewton** [Soostone/retry#28]
I haven't yet been able to profile this and track it down precisely.  But based on circumstantial evidence it looks like retry was responsible for 100% CPU usage as discussed in this commit:

   https://github.com/rrnewton/stack/commit/52f88b58fa45714f0234bf038177daf1a5d03ebc

Which is part of this PR:

   https://github.com/commercialhaskell/stack/pull/670

As seen in the commit, the retry policy was:

```Haskell
retrying (capDelay (500*1000) (exponentialBackoff 1000))
```

So it should have rapidly backed off to polling every 500ms.  And the "try to grab lock" action running at 2hz should have been very cheap.  But what I instead saw was that CPU usage while it was in this polling state would hang out at 2-3% for a few seconds, and then spike to 100+% CPU (all of one core), and stay there indefinitely.

This could be related to some kind of GHC runtime problem.  In particular, Stack uses `+RTS -N` (https://github.com/commercialhaskell/stack/issues/680), and something may be going wrong with the extra HECs.

But in any case, switching stack to a blocking operation definitely got rid of the problem.

**Comment by ozataman** [comment]
That's rather unexpected. You start with a 1ms base delay, so I would expect a burst initially, but it should then slow down. However, IIRC capDelay will limit the max time of delay but it will not limit the number of retries. So it makes sense that this eventually keeps forever retrying at 500ms delay. 

I'll look into the codepath in retry to see if there's anything out of place. We do have tests confirming the policies though, so not sure what could be causing it...

**Comment by rrnewton** [comment]
stack is easy to build and run btw, so it may be a good reproducer to just check out that particular commit of stack.

**Comment by MichaelXavier** [comment]
Tried to reproduce this and couldn't. Going to close it until we get a repro case.

Here's what I tried

`Main.hs`
```
module Main
    ( main
    ) where

import Control.Retry

main :: IO ()
main = work
  where policy = capDelay (500*1000) (exponentialBackoff 1000)
        work = retrying policy (\_ _ -> return True) (return ())
```

Ran with
`cabal exec -- ghc --make -rtsopts -threaded Main.hs && ./Main +RTS -N`

CPU was at a minimal use. I also had a variant that printed at retry intervals and it clearly ramped up to a 500ms delay and then stayed there. This was on 64 bit linux, Ubuntu 14.04.
